Replace ButtonBar.bar method with ButtonBarTheme

darrenaustin opened the pull request:
## Description

This PR implements a proposed change to deprecate `ButtonBar.bar` and replace it with a new `ButtonBarTheme` to configure `ButtonBar` widgets.

### Summary of changes:
- Adds new `ButtonBarTheme` inherited widget and `ButtonBarThemeData` configuration class.
- Adds new `ThemeData.buttonBarTheme` field that can be used to set the defaults for all `ButtonBar` widgets in the app.
- Updates `ButtonBar` to use the surrounding `ButtonBarTheme` to configure itself and override the surrounding `ButtonTheme` for buttons contained within itself.
- Deprecates `ButtonTheme.bar`.
- Changes the default `ButtonTextTheme` to `primary` for buttons in an `ButtonBar` to match the Material Design specification and make the button text easier to read. This change shown with the baseline Material theme:

| ButtonTextTheme.accent  | ButtonTextTheme.primary |
| ------------- | ------------- |
| ![accent](https://user-images.githubusercontent.com/19588/62242104-51779f00-b38f-11e9-98b5-b7ca1c5b886c.png) | ![primary](https://user-images.githubusercontent.com/19588/62242120-563c5300-b38f-11e9-9950-ee4c333f7a7f.png) |

These changes will allow us to fix this accessibility problem with button text as well as providing a more flexible way to configure `ButtonBar` widgets across the entire application (including internal usages like `Dialog`).

## Related Issues

Fixes #37333
Fixes #19946
Fixes #20585

## Tests

I added unit tests for everything in the new `ButtonBarTheme` classes, and updated the unit tests for `ButtonBar`.

## Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require Flutter developers to manually update their apps to accommodate your change?

- [x] Yes, this is a breaking change ([announcement email](https://groups.google.com/forum/#!topic/flutter-announce/Igg0DuIO6Zg))

### How to update your code

With this proposed change, if you have code that is just wrapping a `ButtonBar` in `ButtonTheme.bar`:

```dart
ButtonTheme.bar(
  textTheme: ButtonTextTheme.accent,
  child: ButtonBar(...)
)
```

you can simply remove the `ButtonTheme.bar` wrapper and move the parameters to the `ButtonBar` constructor:

```dart
ButtonBar(
  buttonTextTheme: ButtonTextTheme.accent,
  ...
)
```

If you would like to restore the default `ButtonTextTheme` used by all `ButtonBar` widgets to `ButtonTextTheme.accent`, you can just add the following to your app's theme:

```dart
MaterialApp(
  theme: ThemeData(
    buttonBarTheme: ButtonBarThemeData(
      buttonTextTheme: ButtonTextTheme.accent,
    ),
  ),
  ...
)
```

<!-- Links -->
[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

darrenaustin committed:
Added new ButtonBarTheme to replace the deprecated ButtonTheme.bar method.

rami-a commented on packages/flutter/lib/src/material/button_bar.dart:
```diff
@@ -12,12 +13,18 @@ import 'raised_button.dart';
 ///
+/// The [ButtonBar] can be configured with a [ButtonBarTheme]. For any property
+/// not set on the [ButtonBar] directly it will look for a surrounding
+/// [ButtonBarTheme] and use its properties. The button values in in
```
An extra `in`

rami-a commented on packages/flutter/lib/src/material/button_bar.dart:
```diff
@@ -26,43 +33,117 @@ import 'raised_button.dart';
   /// How the children should be placed along the horizontal axis.
+  ///
+  /// If `null` then it will use [ButtonBarTheme.alignment]. If that is `null`,
+  /// it will default to [MainAxisAlignment.end]
```
Period at end of sentence.

rami-a commented on packages/flutter/lib/src/material/button_bar.dart:
```diff
@@ -26,43 +33,117 @@ import 'raised_button.dart';
   /// How much horizontal space is available. See [Row.mainAxisSize].
+  ///
+  /// If `null` then it will use the surrounding [ButtonBarTheme.mainAxisSize].
+  /// If that is `null`, it will default to [MainAxisSize.max]
```
Period at end of sentence.

rami-a commented on packages/flutter/lib/src/material/button_bar.dart:
```diff
@@ -26,43 +33,117 @@ import 'raised_button.dart';
+  /// base colors, size, internal padding and shape.
+  ///
+  /// If `null` then it will use the surrounding [ButtonBarTheme.buttonTextTheme].
+  /// If that is `null`, it will default to [ButtonTextTheme.primary]
```
Period at end of sentence.

rami-a commented on packages/flutter/lib/src/material/button_bar_theme.dart:
```diff
@@ -0,0 +1,238 @@
+  /// This will override the surrounding [ButtonTheme.textTheme] setting
+  /// for buttons contained in the [ButtonBar].
+  ///
+  /// Despite the name, this property is not a [TextTheme], its value is not a
```
I've been confused by this before, good call on explaining it.

HansMuller replied:
NICE

rami-a commented on packages/flutter/lib/src/material/button_bar_theme.dart:
```diff
@@ -0,0 +1,238 @@
+    );
+  }
+
+  /// Linearly interpolate between two tooltip themes.
```
s/tooltip/buttonbar

darrenaustin replied:
D'oh! thx.

rami-a commented on packages/flutter/lib/src/material/scaffold.dart:
```diff
@@ -2140,12 +2139,10 @@ class ScaffoldState extends State<Scaffold> with TickerProviderStateMixin {
-                child: ButtonBar(
-                  children: widget.persistentFooterButtons,
-                ),
+            child: SafeArea(
```
Hmm I know you didn't add this but the double safe area seems a bit strange to me, is this necessary?

darrenaustin replied:
yeah, not sure what was up with that. I was just mechanically removing `ButtonTheme.bar` but looking at it now that is really odd. I removed it and it doesn't seem to break anything.

rami-a committed:
[Material] Create material Banner component (#36880)

This PR creates a new material widget for the Banner component. This includes a theme as well. This widget can be dropped into any application, ideally at the top of a listview or scrollview.

(cherry picked from commit 35b6d668e197035ea2b57adb99041542982e8be0)

Removed the use of ButtonTheme.bar in the Banner implementation.

rami-a commented:
@darrenaustin I just landed https://github.com/flutter/flutter/pull/36880, could you update your PR to include the changes for ButtonBar in MaterialBanner too, if its not too much trouble?

darrenaustin committed:
Responding to PR feedback.

rami-a reviewed (approved):
This looks great, thanks @darrenaustin

darrenaustin commented:
> @darrenaustin I just landed #36880, could you update your PR to include the changes for ButtonBar in MaterialBanner too, if its not too much trouble?

No problem at all. In fact I need to anyway, as the two themes added to `ThemeData` collide.

HansMuller commented on packages/flutter/lib/src/material/button_bar.dart:
```diff
@@ -26,43 +33,117 @@ import 'raised_button.dart';
 
   /// How the children should be placed along the horizontal axis.
+  ///
+  /// If `null` then it will use [ButtonBarTheme.alignment]. If that is `null`,
```
We don't usually put null in back ticks.

HansMuller commented on packages/flutter/lib/src/material/button_bar.dart:
```diff
@@ -26,43 +33,117 @@ import 'raised_button.dart';
+  /// minimum width.
+  ///
+  /// If `null` then it will use the surrounding [ButtonBarTheme.buttonMinWidth].
+  /// If that is `null`, it will default to `64.0` logical pixels.
```
We don't usually put numbers in back ticks either. Or true and false (see below).

darrenaustin replied:
Ok, I thought it was useful to call out constants in the descriptions, but I will take these out.

HansMuller commented on packages/flutter/lib/src/material/button_bar.dart:
```diff
@@ -12,12 +13,18 @@ import 'raised_button.dart';
 /// children are left justified and the last child becomes the leftmost child.
 ///
+/// The [ButtonBar] can be configured with a [ButtonBarTheme]. For any property
+/// not set on the [ButtonBar] directly it will look for a surrounding
```
This is a complicated sentence. Maybe instead of "directly it will look" => the button bar will look

Note also: sometimes it useful to use the link notation for the first reference and then just write out the class name for subsequent references. The font changes implied by [Foo] can make the text harder to read.

darrenaustin replied:
Good to know. Is it just link the first reference in the entire doc string or for each paragraph or sentence?

HansMuller commented on packages/flutter/lib/src/material/button_bar.dart:
```diff
@@ -12,12 +13,18 @@ import 'raised_button.dart';
+/// The [ButtonBar] can be configured with a [ButtonBarTheme]. For any property
+/// not set on the [ButtonBar] directly it will look for a surrounding
+/// [ButtonBarTheme] and use its properties. The button values in
+/// [ButtonBarTheme] will be used to override properties in the surrounding
```
Since it's a little confusing, it might be worth devoting a short paragraph at the end here to the fact that ButtonBar wraps its child with a ButtonTheme whose ButtonThemeData is based on the button bar's parameters and the surrounding (sometimes we say "ambient") ButtonBarTheme.

HansMuller commented on packages/flutter/lib/src/material/button_bar_theme.dart:
```diff
@@ -0,0 +1,238 @@
+/// Defines the visual properties of [ButtonBar] widgets.
+///
+/// Used by [ButtonBarTheme] to control the visual properties of [ButtonBar]
+/// instances in a  widget subtree.
```
extra space after "in a"

HansMuller commented on packages/flutter/lib/src/material/button_bar_theme.dart:
```diff
@@ -0,0 +1,238 @@
+///    bar theme.
+class ButtonBarTheme extends InheritedWidget {
+  /// Constructs a button bar theme that configures all descendent [ButtonBar]
+  /// widgets.
```
Should also say that data must not be null

HansMuller commented on packages/flutter/test/material/button_bar_test.dart:
```diff
@@ -15,105 +15,279 @@ void main() {
+      );
+
+      final Finder child = find.byType(SizedBox);
+      expect(tester.getRect(child).left, 782.0);
```
For tests like this it's helpful to include a comment that explains where the value comes from. In this case 782 = 800 - 10 - [something about the default padding].

HansMuller reviewed (approved):
LGTM

darrenaustin committed:
Merge branch 'master' into button_bar_theme

darrenaustin committed:
Updated documentation from PR review comments.

darrenaustin committed:
Merge branch 'master' into button_bar_theme

darrenaustin committed:
Merge branch 'master' into button_bar_theme
